File: src/leetcode/leetcode_20.cpp
class Solution {
public:
    bool isValid(string s) {
        vector<char> stack;

        int n = s.size();

        for(int i = 0;i < n;i++){
            char ar = s[i];
             switch(ar){
                default:
                    stack.push_back(ar);
                    break;
                case ')':
                    if(!stack.empty() && stack.back()=='(') stack.pop_back();
                    else return false;
                    break;
                case '}':
                    if(!stack.empty() && stack.back()=='{') stack.pop_back();
                    else return false;
                    break;
                case ']':
                    if(!stack.empty() && stack.back()=='[') stack.pop_back();
                    else return false;
                    break;
            }
        }
        return stack.empty();
    }
};

File: src/leetcode/leetcode_20.py
class Solution:
    def isValid(self, s: str) -> bool:
        n = len(s)

        stack = []
        for i in range(0, n):
            ar = s[i]

            if ar=='(' or ar=='[' or ar=='{':
                stack.append(ar)
                continue
            if ar == ')':
                if len(stack)!=0 and stack[-1]=='(':
                    # stack = stack[:-1]
                    stack.pop()
                else:
                    return False
            elif ar == ']':
                if len(stack)!=0 and stack[-1]=='[':
                    # stack = stack[:-1]
                    stack.pop()
                else:
                    return False
            elif ar == '}':
                if len(stack)!=0 and stack[-1]=='{':
                    # stack = stack[:-1]
                    stack.pop()
                else:
                    return False

        return len(stack) == 0